docs/4.x/updates-aka-migrations.md

```diff
@@ -0,0 +1,102 @@
+        $this->migration = $factory;
+    }
+
+    public function getMigrations(Updater $updater)
+    {
+        // many different migrations are available to be used via $this->migration factory
+        $migration1 = $this->migration->db->changeColumnType('log_visit', 'example', 'BOOLEAN NOT NULL');
+        // you can also define custom SQL migrations. If you need to bind parameters, use `->boundSql()`
+        $migration2 = $this->migration->db->sql($sqlQuery = 'SELECT 1');
+
+        return array(
+            // $migration1,
+            // $migration2
+        );
+    }
+
+    public function doUpdate(Updater $updater)
+    {
+        $updater->executeMigrations(__FILE__, $this->getMigrations($updater));
+    }
+}
+```
+
+The `getMigrations()` method returns explicit step by step changes to make to the system. Separating them like
+this allows us to display the list of changes that will be made to the user in case the user would like to do
+them manually, one at a time (useful for large instances).
+
+The `doUpdate()` method performs the entire update. It must call `$updater->executeMigrations(...)` as above if there are
+individual migrations defined. It can also include other update logic that doesn't have an associated migration,
+but it is encouraged to find some way to put them in a migration, since otherwise the user wouldn't be able
+to apply them manually.
+
+If you do create a custom migration, it is required that you make it **idempotent**. That is to say, running the
+update multiple times should change the system only once and should not result in an error.
+
+**Steps to creating a new Update**
+
+To create a new Update class, do the following:
+
+1. Check the version in core/Version.php, and bump it. If it's set to `4.0.0`, for example, set it to `4.0.0-b1`. If it's set to `4.0.0-b1`, set it to `4.0.0-b2`.
+   If it's set to `4.0.0-rc1`, set it to `4.0.0-rc2`. The version you bump it to must always be a beta or release candidate version, since the product
+   owner is the only one who releases whole new versions.
+2. Run the `./console generate:update` command and specify `core`.
+3. In the `core/Updates` folder there will now be a new Update with the version you bumped Version.php to. Fill it with the update
+   logic you need.
+
+## How Updates Are Run
+
+On every request Matomo will check if the current version of Matomo (stored in the `version_core` option) is less than the version hardcoded in Version.php.
+If this is true, then the code was updated, but not the database. This triggers the database update workflow.
+
+The update workflow looks for every update file in the `Updates` folder that has a version greater than the current version and less than or equal to the version in Version.php (using `version_compare()`). It runs each in order, one after the other. Once done, the `version_core` option is updated.
+
+The updater will also automatically make changes to tracking dimensions if their definition changes.
```

Comment:
Can we maybe move the content from https://developer.matomo.org/guides/contributing-to-piwik-core#dealing-with-matomo-updates in here somewhere too? And then keep the title in https://developer.matomo.org/guides/contributing-to-piwik-core#dealing-with-matomo-updates but change the paragraph to only link to this guide similar to 
![image](https://user-images.githubusercontent.com/273120/113198302-4393c300-92c2-11eb-9d3b-a03a8689ca7b.png)

Reply:
Those docs are more about dealing with the need to update while developing Matomo, while this guide is about how they are written and how they generally work. I can't think of a good place to put those two settings in this guide, so I just linked to it in contributing-to-piwik-core

Reply:
👍